### ISSUE TITLE
SQS receive-message delays/sleeps

### ISSUE THREAD

**Issue by lobster1234** [atlassian/localstack#52]
I am observing that calls to receive-message on SQS seem to take quite a long time.

Here are the details - 

```
C02STG51GTFM:localstack mpandit$  aws --endpoint-url=http://localhost:4576 sqs send-message --queue-url http://localhost:4576/123456789012/test_queue --message-body 'Test Message!'
{
    "MD5OfMessageBody": "df69267381a60e476252c989db9ac8ad", 
    "MessageId": "dc8cd4d1-53ec-b8a3-63d2-457fb4ac4019"
}
C02STG51GTFM:localstack mpandit$ time aws --endpoint-url=http://localhost:4576 sqs receive-message --queue-url http://localhost:4576/123456789012/test_queue
{
    "Messages": [
        {
            "Body": "Test Message!", 
            "Attributes": {
                "ApproximateFirstReceiveTimestamp": "1.49138201347e+12", 
                "SenderId": "AIDAIT2UOQQY3AUEKVGXU", 
                "ApproximateReceiveCount": "2", 
                "SentTimestamp": "1.49138201346e+12"
            }, 
            "ReceiptHandle": "guoazsjkomdujlaegxlqzauynkketuxvcomewlvhbdkjyifrgirnobfxifjvkuztoeeuwkkmnxkbelyiyhbrldqgfpydmywtmoykjrimkhiezejsbxvotivjvinudnxxyjltnaoecaduxplgpyrigeqwgsdbhoqtsotysguttwcpmzonqmhnemuuc", 
            "MD5OfBody": "df69267381a60e476252c989db9ac8ad", 
            "MessageId": "dc8cd4d1-53ec-b8a3-63d2-457fb4ac4019"
        }
    ]
}

real	0m26.498s
user	0m0.309s
sys	0m0.168s
C02STG51GTFM:localstack mpandit$ time aws --endpoint-url=http://localhost:4576 sqs receive-message --queue-url http://localhost:4576/123456789012/test_queue
{
    "Messages": [
        {
            "Body": "Test Message!", 
            "Attributes": {
                "ApproximateFirstReceiveTimestamp": "1.49138201347e+12", 
                "SenderId": "AIDAIT2UOQQY3AUEKVGXU", 
                "ApproximateReceiveCount": "3", 
                "SentTimestamp": "1.49138201346e+12"
            }, 
            "ReceiptHandle": "xsxjduahofaqqisgjybfpfzzqgceiaiduscqnygkhalvzzfqrsltbbavhopjbqptrazwdjriuahtlswdezpnlflvxqaualqmabwiobecdttrzwpnjjvzekmbbgnjuvwvfwwoipeqbcctlcujtlzvxdvoqszmmhkrrxfhdfuzbfgebxugcihezefjv", 
            "MD5OfBody": "df69267381a60e476252c989db9ac8ad", 
            "MessageId": "dc8cd4d1-53ec-b8a3-63d2-457fb4ac4019"
        }
    ]
}

real	0m19.425s
user	0m0.281s
sys	0m0.162s
```

**Comment by whummer** [comment]
Thanks for reporting. Do you see this behavior when starting LocalStack in Docker (`make run-docker`) or on the local machine (`make infra`)?

**Comment by lobster1234** [comment]
```make-infra```.

**Comment by whummer** [comment]
Turns out that this is actually the expected behavior. The SQS API defines a **message visibility timeout**, which is used to deal with "in flight" messages and to give the consumers enough time to process and then finally delete the message from the queue. Note that messages stay available in the queue as long as they have not been explicitly deleted by one of the consumers. For details see here: http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ChangeMessageVisibility.html

The AWS API documentation mentions that the message visibility timeout defaults to 5 minutes, whereas in the local test environment it defaults to 30 seconds (which makes it more useful for testing purposes). See the underlying implementation for more details: https://github.com/spulec/moto/blob/042524fe806e59e61fa252bad4647ad8f4231af4/moto/sqs/models.py#L272

In other words, if you run the `aws sqs receive-message` you posted above, then wait 30 seconds, then run the command again, the result will be returned immediately. Alternatively, use `aws sqs delete-message ...` to delete a message from the queue after it has been processed by your consumer.

**Comment by lobster1234** [comment]
I get the visibility timeout, but that should not block the caller, but "hide" that message for 30 seconds to avoid multiple delivery. In fact I was hoping it would honor the visibility timeout but I could not verify due to the blocking call.

**Comment by whummer** [comment]
Oh right, my bad, I missed that part. I just gave it another try and tested `receive-message` with `--wait-time-seconds 1` parameter, but the parameter is not processed properly.

I'm going to raise a PR against `moto` because there is a bug in their polling loop (it just keeps re-trying and never gets to the point where it should break if `unix_time() > polling_end`): https://github.com/spulec/moto/blob/042524fe806e59e61fa252bad4647ad8f4231af4/moto/sqs/models.py#L292

**Comment by lobster1234** [comment]
Thanks so much 👍

**Comment by whummer** [comment]
Should be fixed in version `0.3.9`. Please re-open if the problem persists.
